polyply/src/topology.py

```diff
@@ -216,14 +224,55 @@ def gen_pairs(self):
+                                                {"nb1": nb1, "nb2": nb2}})
 
         for atom_type in self.atom_types:
             if frozenset([atom_type, atom_type]) not in self.nonbond_params:
                 nb1 = self.atom_types[atom_type]["nb1"]
                 nb2 = self.atom_types[atom_type]["nb2"]
                 self.nonbond_params.update({frozenset([atom_type, atom_type]):
-                                           {"nb1": nb1, "nb2": nb2}})
+                                            {"nb1": nb1, "nb2": nb2}})
+
+    def gen_bonded_interactions(self):
+        """
+        Check for each interaction if there is
+        no parameter for an interaction if that
+        parameter is defined in the bonded directive
+        of the topology.
+        """
+        for block in self.force_field.blocks.values():
+            for inter_type, interactions in block.interactions.items():
+                if inter_type in ["pairs", "exclusions", "virtual_sitesn"]:
+                    continue
+                for interaction in interactions:
+                    if len(interaction.parameters) == 1:
+                        atoms = tuple(block.nodes[node]["atype"] for node in interaction.atoms)
+
+                        if atoms in self.types[inter_type]:
+                            new_params, meta = self.types[inter_type][atoms]
+                        elif atoms[::-1] in self.types[inter_type]:
+                            new_params, meta = self.types[inter_type][atoms[::-1]]
+                        elif inter_type in "dihedrals" and\
+                             ("X", atoms[1], atoms[2], "X") in self.types[inter_type]:
+                            new_params, meta = self.types[inter_type][("X", atoms[1], atoms[2], "X")]
+                        elif inter_type == "dihedrals" and\
+                             ("X", atoms[2], atoms[1], "X") in self.types[inter_type]:
+                            new_params, meta = self.types[inter_type][("X", atoms[2], atoms[1], "X")]
+                        elif inter_type in "dihedrals" and\
+                             (atoms[0], "X", "X", atoms[3]) in self.types[inter_type]:
+                            new_params, meta = self.types[inter_type][(atoms[0], "X", "X", atoms[3])]
+                        elif inter_type == "dihedrals" and\
+                             (atoms[3], "X", "X", atoms[0]) in self.types[inter_type]:
+                            new_params, meta = self.types[inter_type][(atoms[3], "X", "X", atoms[0])]
+                        else:
+                            msg=("In section {} interaction of atoms {} has no corresponding bonded"
+                                 "type.")
+                            atoms = " ".join(list(map(lambda x: str(x), interaction.atoms)))
+                            raise OSError(msg.format(inter_type, atoms))
+
+                        interaction.parameters[:] = new_params[:]
```

Comment:
See also https://docs.python.org/3.8/library/collections.html#collections.somenamedtuple._replace

Reply:
this returns a new instance but I want to keep the old one and just overwrite the parameters

Reply:
This on the other hand will only work when the parameters are lists, which is not defined. See also https://github.com/marrink-lab/vermouth-martinize/issues/240